### PR TITLE
Upgrade fastapi to 0.114.1

### PR DESCRIPTION
**Pull request by rralcala**
Also resolves some deprecation warnings from sqlalchemy and fix a test.

### CODE REVIEW

File: tests/service.py [hunk 1]
```diff
@@ -322,7 +322,7 @@ def test_delete_provider(client):
 @pytest.mark.dependency(depends=["test_delete_provider"])
 def test_empty_measurements(client):
     query = {
-        "start": "1984-04-24T00:00:00",
+        "start": "2984-04-24T00:00:00",
```
Comment (tchx84) [inline tests/service.py:325]: is this change still necessary? If not, let's just revert it.

File: tests/service.py [hunk 2]
```diff
@@ -75,7 +75,7 @@
         "co2": 200.0,
         "longitude": -57.521369,
         "latitude": -25.194156,
-        "recorded": "2020-10-24T20:47:57.370721+00:00",
+        "recorded": "2020-10-24T20:47:57.370721Z",
```
Comment (tchx84) [inline tests/service.py:78]: curious, why is this change needed?

File: app/schemas.py [hunk 3]
```diff
@@ -125,36 +125,37 @@ def to_orm(self, provider):
 
     class Config:
         orm_mode = True
+        from_attributes = True
 
 
 @dataclass
 class QueryParams:
-    source: str = Query(
+    source: Optional[str] = Query(
```
Comment (tchx84) [inline app/schemas.py:133]: curious, why these explicit "optional" are needed?

File: app/reports.py [hunk 4]
```diff
@@ -71,9 +71,11 @@ def get_quality(source):
         breakpoint_index = CONCENTRATIONS.index(concentration)
         breakpoint = BREAKPOINTS[breakpoint_index]
 
-        index = (
-            (breakpoint[1] - breakpoint[0]) / (concentration[1] - concentration[0])
-        ) * (source.pm2dot5_average - concentration[0]) + breakpoint[0]
+        index = int(
```
Comment (tchx84) [inline app/reports.py:74]: curious why was this needed? It isn't an actual array index.

File: README.md [hunk 5]
```diff
@@ -4,6 +4,8 @@ LINKA is an open source service to record air quality data from community sensor
 
 ## Install
 
+Python 3.12 is not yet supported
```
Comment (tchx84) [inline README.md:7]: Let's move this as a proper note at the bottom of the subsection, like

> [!NOTE]  
> Python 3.12 is not yet supported.

```
> [!NOTE]  
> Python 3.12 is not yet supported.
```

**Review by tchx84 — CHANGES_REQUESTED** [review]
Thanks @rralcala , just minor comments and a few questions for me curiosity.

File: tests/service.py (hunk 1 above)
Comment (rralcala) [reply tests/service.py:325]:
```
test_empty_measurements(client):
    query = {
        "start": "2984-04-24T00:00:00",
    }
```
```
measurements = [
    {
        "sensor": "aqi",
        "source": "aqi",
        "description": "aqi",
        "version": None,
        "pm1dot0": 26.4,
        "pm2dot5": 26.4,
        "pm10": 26.4,
        "humidity": 100.0,
        "temperature": 134.0,
        "pressure": 1084.0,
        "co2": 1000.0,
        "longitude": -57.521369,
        "latitude": -25.194156,
        "recorded": "2020-10-24T20:47:57.370721Z",
    },
    {
        "sensor": "nullable",
        "source": "nullable",
        "description": "nullable",
        "version": "nullable",
        "pm1dot0": None,
        "pm2dot5": None,
        "pm10": None,
        "humidity": None,
        "temperature": None,
        "pressure": None,
        "co2": None,
        "longitude": -57.521369,
        "latitude": -25.194156,
        "recorded": "2020-10-24T20:47:57.370721Z",
    },
    {
        "sensor": "test",
        "source": "test",
        "description": None,
        "version": None,
        "pm1dot0": 0.0,
        "pm2dot5": 0.0,
        "pm10": 0.0,
        "humidity": 1.0,
        "temperature": -89.2,
        "pressure": 870.0,
        "co2": 200.0,
        "longitude": -57.521369,
        "latitude": -25.194156,
        "recorded": "2020-10-24T20:47:57.370721Z",
    },
]
```
If I revert, will the result be empty?

File: app/reports.py (hunk 4 above)
Comment (rralcala) [reply app/reports.py:74]: The casting to int is because int is expected.

File: app/schemas.py (hunk 3 above)
Comment (rralcala) [reply app/schemas.py:133]: Because some tests send None in the source field?

File: tests/service.py (hunk 2 above)
Comment (rralcala) [reply tests/service.py:78]: Output changed somewhere after the upgrade, '+00:00' is equivalent to Z anyway.

**Comment by rralcala** [comment]
The answer to most of the comments is: Fastapi pulls a stricter version of pydantic, so issues that weren't raised previously are being raised now.

File: tests/service.py (hunk 1 above)
Comment (tchx84) [reply tests/service.py:325]: Yes, because the previous test is deleting the provider and should delete its measurements as well, and that's we're testing :)

Comment (rralcala) [reply tests/service.py]: I did miss this line
`@pytest.mark.dependency(depends=["test_delete_provider"])`
:)

File: requirements.txt [hunk 6]
```diff
@@ -9,8 +9,9 @@ black==23.1.0
...
-asyncpg==0.27.0
-psycopg2-binary==2.9.5
+asyncpg==0.29.0
+psycopg2-binary==2.9.9
 alembic==1.10.2
 geopy==2.3.0
-pytest-dependency==0.5.1
\ No newline at end of file
+pytest-dependency==0.5.1
+pydantic==2.9.1
```
Comment (tchx84) [inline requirements.txt:17]: I missed this one, I think there's no need to require pydantic explicitly, since fastapi should pull the right version...

**Review by tchx84 — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 6 above)
Comment (rralcala) [reply requirements.txt:17]: That's because on my arm Mac (don't know which one is the problem) pulls a different version with different errors, so locking in the newest version for now helps getting a consistent set of issues.

I guess that when all the remaining set of dependencies + python can be upgraded some of these will go away.

**Review by tchx84 — APPROVED** [review]
